src/cephadm/cephadm

```diff
@@ -4415,82 +4422,162 @@ def is_ipv6(address):
         return False
 
 
-def prepare_mon_addresses(
-    ctx: CephadmContext
-) -> Tuple[str, bool, Optional[str]]:
+def ip_in_subnets(ip_addr: str, subnets: str) -> bool:
+    """Determine if the ip_addr belongs to any of the subnets list."""
+    subnet_list = [x.strip() for x in subnets.split(',')]
+    for subnet in subnet_list:
+        if ipaddress.ip_address(ip_addr) in ipaddress.ip_network(subnet):
+            return True
+    return False
+
+
+def parse_mon_addrv(addrv_arg: str) -> List[EndPoint]:
+    """Parse mon-addrv param into a list of mon end points."""
     r = re.compile(r':(\d+)$')
-    base_ip = ''
+    addrv_args = []
+    addr_arg = addrv_arg
+    if addr_arg[0] != '[' or addr_arg[-1] != ']':
+        raise Error(f'--mon-addrv value {addr_arg} must use square backets')
+
+    for addr in addr_arg[1: -1].split(','):
+        hasport = r.findall(addr)
+        if not hasport:
+            raise Error(f'--mon-addrv value {addr_arg} must include port number')
+        port_str = hasport[0]
+        addr = re.sub(r'^v\d+:', '', addr)  # strip off v1: or v2: prefix
+        base_ip = addr[0:-(len(port_str)) - 1]
+        addrv_args.append(EndPoint(base_ip, int(port_str)))
+
+    return addrv_args
+
+
+def parse_mon_ip(mon_ip: str) -> List[EndPoint]:
+    """Parse mon-ip param into a list of mon end points."""
+    r = re.compile(r':(\d+)$')
+    addrv_args = []
+    hasport = r.findall(mon_ip)
+    if hasport:
+        port_str = hasport[0]
+        base_ip = mon_ip[0:-(len(port_str)) - 1]
+        addrv_args.append(EndPoint(base_ip, int(port_str)))
+    else:
+        # Use fixd ports for ceph monitor
+        addrv_args.append(EndPoint(mon_ip, 3300))
+        addrv_args.append(EndPoint(mon_ip, 6789))
+
+    return addrv_args
+
+
+def build_addrv_params(addrv: List[EndPoint]) -> str:
+    """Convert mon end points (ip:port) into the format: [v[1|2]:ip:port1]"""
+    addr_arg = ''
+    if len(addrv) == 1:
+        ip = addrv[0].ip
+        port = addrv[0].port
+        if port == 6789:
+            addr_arg = f'[v1:{ip}]'
+        elif port == 3300:
+            addr_arg = f'[v2:{ip}]'
+        else:
+            logger.warning(f'Using msgr2 protocol for unrecognized port {port}')
+            addr_arg = f'[v2:{ip}]'
+    elif len(addrv) == 2:
+        ip = addrv[0].ip
+        addr_arg = f'[v2:{ip}:3300,v1:{ip}:6789]'
```

Comment:
I know the way this is currently being used it always gets the result of `parse_mon_ip` so we know if the length is two the ports are static, but do we want to verify it's the expected ports anyhow since we've now moved this into its own function just in case we try to use it elsewhere in the future?

Reply:
I've re-written the code in a more concise and generic code. Please, have a look and let me know your your thoughts.

Reply:
I like the generic version better :+1:. We should probably still error though if then length is > 2 though, right? Given this is just the mon addrv for a single mon we're creating it shouldn't have more than 2 entries.

Reply:
Done.